src/action-sheets/index.js

```diff
@@ -179,6 +185,85 @@ const markTopicAsRead = {
   },
 };
 
+const markAsUnreadFromMessage = {
+  title: 'Mark as unread from here',
+  errorMessage: 'Failed to mark as unread',
+  action: async ({
+    auth,
+    zulipFeatureLevel,
+    setDoNotMarkMessagesAsRead,
+    narrow,
+    message,
+    _,
+    allUsersById,
+    streams,
+  }) => {
+    setDoNotMarkMessagesAsRead(true);
+
+    // TODO(server-6.0): Simplify this away.  (This is for api.updateMessageFlagsForNarrow.)
+    invariant(zulipFeatureLevel >= 155, 'markAsUnreadFromMessage should be called only at FL 155');
+
+    const apiNarrow = apiNarrowOfNarrow(narrow, allUsersById, streams);
+    // Ideally we might add another element to the API narrow like:
+    //   { negated: true, operator: 'is', operand: 'unread' },
+    // That way we'd skip acting on already-unread messages.
+    //
+    // Unfortunately that *also* skips any messages we lack a UserMessage
+    // record for -- which means stream messages where we weren't subscribed
+    // when the message was sent (and haven't subsequently created such a
+    // record, e.g. by starring the message.)
+
+    let response_count = 0;
+    let affected_count = 0;
+    const args = {
+      anchor: message.id,
+      include_anchor: true,
+      num_before: 0,
+      num_after: 5000,
+      op: 'remove',
+      flag: 'read',
+      narrow: apiNarrow,
+    };
+    while (true) {
+      const result = await api.updateMessageFlagsForNarrow(auth, args);
+      response_count++;
+      affected_count += result.affected_count;
+
+      if (result.found_newest) {
+        if (response_count > 1) {
+          // We previously showed a "working on it…" toast, so say we're done.
+          showToast(_('Marked {numMessages} messages as unread', { numMessages: affected_count }));
+        }
+        return;
+      }
+
+      if (result.last_affected_id === null) {
+        // No messages were in the range of the request.
+        // This should be impossible given that found_newest was false
+        // (and that our num_after was positive.)
+        logging.error('mark-unread: request failed to make progress', result);
+        throw new Error(_('The server sent a malformed response.'));
```

Comment:
I notice this is translated, but I don't expect this message will ever reach the user. I think any error thrown will cause `markAsUnreadFromMessage.errorMessage` to be shown, in translation, in an error alert. That string is "Failed to mark as unread".

Reply:
It shows both -- that latter string is the title of the alert, and the error's own message is the body. I saw this in manual testing with an artificial error, and here's the code in `makeButtonCallback`:
```js
      try {
        await pressedButton.action(args);
      } catch (err) {
        // […]
        Alert.alert(args._(pressedButton.errorMessage), err.message);
      }
```

Reply:
Oh OK, I missed that. Thanks!